src/Mod/Sketcher/Gui/DrawSketchHandlerRectangle.h

```diff
@@ -891,9 +892,17 @@ class DrawSketchHandlerRectangle: public DrawSketchHandlerRectangleBase
                                           Sketcher::PointPos::none,
                                           firstCurve + 7);  // NOLINT
 
-                    if (fabs(thickness) > Precision::Confusion()) {
-                        if (radiusFrame
-                            < Precision::Confusion()) {  // case inner rectangle is normal rectangle
+                    if (thicknessNotNull) {
+                        // There are 3 cases possible:
+                        // 1 - Thickness is negative and -thickness == radius.
+                        // In this case the inner rectangle is a normal rectangle and its corner
+                        // match the centers of the outer arcs.
+                        // 2 - Thickness is negative and  radius < -thickness.
+                        // In this case it's a normal rectangle but we need construction
+                        // lines to constraint it.

```

Comment:
If possible, I'd suggest making helper functions to create the actual rectangles, then calling them from here, and once they are made adding any constraints between them. For example 
```
if (...) {
    firstCurveOfFirstRect = createRoundedRectangle(corner1, corner2, radius);
}
...
if (...) { // case 1
    firstCurveOfSecondRect = createNormalRectangle(corner1, corner2);
}
...
if (...) { // case 1 inter-rectangle constraints
    ...
    addToShapeConstraints(Sketcher::Coincident,
                          firstCurveOfFirstRect + 4,  // NOLINT
                          Sketcher::PointPos::mid,
                          firstCurveOfSecondRect,  // NOLINT
                          Sketcher::PointPos::start);
    ...
}
```
Of course, treat this as a pseudo-code. I am not aware of this DSH.

Reply:
Initially it was as you describe : case 1 was in case 2. But it was failing and throwing error (and therefor this PR) because the case 2 creates construction lines, and in the case of 1, these lines were bad (start and endpoint coincident).

This DSH is a beast and would benefit from a refactoring to be sure. But I would say that it's out of the scope of this PR.

Reply:
> Initially it was as you describe : case 1 was in case 2. But it was failing and throwing error (and therefor this PR) because the case 2 creates construction lines, and in the case of 1, these lines were bad (start and endpoint coincident).

I see. Those lines seem to be a slightly complicated method of constraining the frame to a single uniform thickness. 

> This DSH is a beast and would benefit from a refactoring to be sure. But I would say that it's out of the scope of this PR.

It will only be a bigger headache in the future. At present I cannot at-a-glance understand what either the present or the new code is doing, and the newer code is only more complex. I would highly recommend at least "extracting the method" to create individual rectangles.